File: binary_search/rotated_array/find_min_in_rotated_sorted_array.rb
def find_min(nums)
    l, r = 0, nums.size - 1

    while l < r
        return nums[l] if nums[l] < nums[r]

        mid = (l + r) / 2
        nums[mid] > nums[r] ? l = mid + 1 : r = mid
    end

    nums[l]
end


# [3,4,5,1,2]

# 5,6,7,1,2,3,4


# 153. Find Minimum in Rotated Sorted Array
# https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/description/

# 1. Find inflection point
# 2. Find an array such that first element is smaller than the last
# 3. Keep eliminating half until we are left 
# Time: O(log(n))
# Space: O(1)

require 'test/unit'
extend Test::Unit::Assertions

assert_equal(find_min([3,4,5,1,2]), 1)
assert_equal(find_min([4,5,6,7,0,1,2]), 0)
assert_equal(find_min([1]), 1)
assert_equal(find_min([1, 1]), 1)
assert_equal(find_min([]), -1)

assert_equal(find_min_simple([3,4,5,1,2]), 1)
assert_equal(find_min_simple([4,5,6,7,0,1,2]), 0)
assert_equal(find_min_simple([1]), 1)
assert_equal(find_min_simple([]), nil)

File: sandbox/find_min_in_rotated_sorted_array.py
class Solution:
    def findMin(self, nums):
        l, r = 0, len(nums) - 1

        while l < r:
            if nums[l] < nums[r]: return nums[l]
                
            mid = (l + r) // 2
            if nums[mid] > nums[r]:
                l = mid + 1
            else:
                r = mid

        return nums[l]

# [4,5,6,7,0,1,2]


# [1,2,3,4,5,6]
# [2,3,4,5,6,1,]


# (i.e.,  [0,1,2,4,5,6,7] might become  ).
#         [6,7,0,1,2,4,5] - 1


class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # If the list has just one element then return that element.
        if len(nums) == 1:
            return nums[0]

        l = 0
        r = len(nums) - 1

        if nums[r] > nums[0]:
            return nums[0]

        while l <= r:
            mid = l + (r - l) / 2

            if nums[mid] > nums[mid + 1]:
                return nums[mid + 1]

            if nums[mid - 1] > nums[mid]:
                return nums[mid]
            if nums[mid] > nums[0]:
                l = mid + 1
            else:
                r = mid - 1
